### PR TITLE
Publish container images

### PR DESCRIPTION
**Pull request by aaronmondal**
Introduce signed, fully reproducible container images that are created
on pushes to main and published via GitHub packages.

- - -
This change is [Reviewable](https://reviewable.io/reviews/TraceMachina/native-link/443)

### CODE REVIEW

**Review by aaronmondal — COMMENTED** [review]
+@allada +@MarcusSorealheis +@adam-singer +@blakehatch 

You can ignore the first commit as that'll be "rebased away" after #442.

> *[Reviewable](https://reviewable.io/reviews/TraceMachina/native-link/443#-:-NkrPssGCO_7zEwXPjWP:b-k7jsok)* status: 0 of 8 files reviewed, all discussions resolved (waiting on @adam-singer, @allada, @blakehatch, and @MarcusSorealheis)

File: README.md [hunk 1]
```diff
@@ -2,132 +2,131 @@
...
+Native link is an extremely (blazingly?) fast and efficient build cache and
+remote executor for systems that communicate using the [Remote execution
+protocol](https://github.com/bazelbuild/remote-apis/blob/main/build/bazel/remote/execution/v2/remote_execution.proto) such as [Bazel](https://bazel.build), [Buck2](https://buck2.build) and [Goma](https://chromium.googlesource.com/infra/goma/client/).
 
-The goals of this project are:
-1. Stability - Things should work out of the box as expected
-2. Efficiency - Don't waste time on inefficiencies &amp; low resource usage
-3. Tested - Components should have plenty of tests &amp; each bug should be regression tested
-4. Customers First - Design choices should be optimized for what customers want
+Supports Unix based operating systems and Windows.
```
Comment (MarcusSorealheis) [inline README.md]: Unix-based

**Review by MarcusSorealheis — APPROVED** [review]
LGTM

**Review by allada — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/native-link/443#-:-NksU8I7DfMaA2ALOiE6:bnfp4nl)* status: 5 of 8 files reviewed, 2 unresolved discussions (waiting on @aaronmondal, @adam-singer, @blakehatch, and @MarcusSorealheis)

___
*[`README.md` line 47 at r2](https://reviewable.io/reviews/TraceMachina/native-link/443#-NksSKDpAWnZzfY-g-WR:-NksSKDpAWnZzfY-g-WS:b1dhqzz) ([raw file](https://github.com/TraceMachina/native-link/blob/7f9309ef2c49b1de9ae35dba142eee04f78bb0ea/README.md#L47)):*
> ```Markdown
> # Verify the signature
> cosign verify ghcr.io/TraceMachina/native-link:${LATEST} \
>     --certificate-identity=https://github.com/TraceMachina/native-link/.github/workflows/image.yaml@refs/heads/main \
> ```

What happens if main is updated but the latest github action has not yet finished uploading? Won't this cause issues?

**Review by aaronmondal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/native-link/443)* status: 4 of 8 files reviewed, 2 unresolved discussions (waiting on @adam-singer, @allada, @blakehatch, and @MarcusSorealheis)

___
*[`README.md` line 9 at r1](https://reviewable.io/reviews/TraceMachina/native-link/443#gh-1414893186:-NkuSR5qEWQrsz-jLEUS:b-896fix) ([raw file](https://github.com/TraceMachina/native-link/blob/bb5db4f3449c0f9ef1ff03074f7aad9dc6147f85/README.md#L9)):*
<details><summary><i>Previously, MarcusSorealheis (Marcus Eagan) wrote…</i></summary><blockquote>

Unix-based
</blockquote></details>

Done.
___
*[`README.md` line 47 at r2](https://reviewable.io/reviews/TraceMachina/native-link/443#-NksSKDpAWnZzfY-g-WR:-NkuSWkj4rsLnxxI570R:b-lqbgr9) ([raw file](https://github.com/TraceMachina/native-link/blob/7f9309ef2c49b1de9ae35dba142eee04f78bb0ea/README.md#L47)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

What happens if main is updated but the latest github action has not yet finished uploading? Won't this cause issues?
</blockquote></details>

I wouldn't say that it's too big of an issue, but it can surely be inconvenient if the image is unexpectedly not ready immediately. I've added a comment above and also added a "production" variant that can be used to fully pins the image.

**Review by aaronmondal — COMMENTED** [review]
> Dismissed @MarcusSorealheis from a discussion.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/native-link/443)* status: 4 of 8 files reviewed, 1 unresolved discussion (waiting on @adam-singer, @allada, @blakehatch, and @MarcusSorealheis)

___
*[`README.md` line 9 at r1](https://reviewable.io/reviews/TraceMachina/native-link/443#gh-1414893186:-NkuY6HZ9p8Rj-04VckZ:bb6un1j) ([raw file](https://github.com/TraceMachina/native-link/blob/bb5db4f3449c0f9ef1ff03074f7aad9dc6147f85/README.md#L9)):*
<details><summary><i>Previously, aaronmondal (Aaron Siddhartha Mondal) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done

**Pull request merged**